### PR TITLE
[ENH] add expanding greedy splitter

### LINKED ISSUE #4874: [ENH] Simplified expanding window splitter anchored at time series end
**Is your feature request related to a problem? Please describe.**
This is not so much a feature request but me sharing a thing I did that works nicely for me, in case others find it useful. Feel free to close.

The idea is that when defining cross validation, most of the time I just care about two things
1. the number of folds (which drives run time and variance of results) and
2. the horizon/number of test items 
 
By default, the splittier should use as much data as possible.

So this is another splitter that takes those two args:
```py
ExpandingGreedySplitter(horizon=7, folds=5)
```

This will create 5 folds, each with 7 days in the test set, arranged such that the test sets in total cover the last 35 days with no overlaps, and each train set is all data before the test set.

(I'm aware that having `horizon` as an int is not the sktime way but this is what works for me and how I think about the size of the validation set in cross validation, so I'm just sharing as is.)

It also takes a `step_length`, which defaults to the same as horizon for simplicity since that's what I want most of the time.

My work flow tends to be to set a low number for folds and horizon in the early stages of grid search to get a ballpark idea of what values work, then increase the folds and horizon to get more refined results. I found with the existing options this was a bit fiddly.

**Describe the solution you'd like**
This is the class. A bit hacky in places, certainly not accounting for all use cases, but again, gets the job done for my uses.
```py
class ExpandingGreedySplitter(BaseSplitter):
    def __init__(self, horizon, folds=5, step_length=None):
        super().__init__()
        self.folds = folds
        self.horizon = horizon
        self.fh = np.arange(horizon) + 1
        self.step_length = step_length or horizon

    def split(self, y):
        y = self._coerce_to_index(y)
        if isinstance(y, pd.MultiIndex):
            dates = y.get_level_values(y.names[-1])
        else:
            dates = y

        if isinstance(dates, pd.PeriodIndex):
            freq = dates.freq
            dates = dates.to_timestamp()
        else:
            freq_str = pd.infer_freq(dates)
            freq = pd.tseries.frequencies.to_offset(freq_str)

        if freq is None:
            print("Couldn't work out frequency, assuming 'Day'")
            freq = pd.tseries.offsets.Day()

        trn_ends = pd.date_range(
            end=dates.max() - freq * self.horizon,
            periods=self.folds,
            freq=freq * self.step_length,
        )

        for trn_end in trn_ends:
            tst_end = trn_end + freq * self.horizon
            trn_indices = np.flatnonzero(dates <= trn_end)
            tst_indices = np.flatnonzero((dates > trn_end) & (dates <= tst_end))
            yield trn_indices, tst_indices
```

This assumes all instances _end_ at the same date but can have any start date (again, works for my case, not sure how often this assumption holds for others).

**Describe alternatives you've considered**
I originally wrote a `make_expanding_window_cv` function that used `ExpandingWindowSplitter` but this didn't work when my instances had different lengths.

### PR DESCRIPTION
**Pull request by davidgilbertson**
#### Reference Issues/PRs

Fixes #4874

#### What does this implement/fix? Explain your changes.
This allows a user to define cross-validation splits in terms of `test_size` and number of folds.

#### Does your contribution introduce a new dependency? If yes, which one?

Nope

#### What should a reviewer concentrate their feedback on?

I wasn't sure about the name of `test_size`, it could also be called `horizon` but I thought that was a bit misleading since it expects an `int` unlike 'forecast horizon'. Open to suggestions.

 - [ ] I just learned from this template about the contributors file, I'll do this tomorrow and update the PR.

#### Did you add any tests for the change?

Yes, but I wasn't sure about the existing machinery like `@pytest.mark.parametrize` or `get_test_params`. If I should be using these please let me know.

#### Any other comments?
My first ever Python PR, be gentle :)

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.

### CODE REVIEW

**Review by fkiraly — COMMENTED** [review]
Nice! Welcome to `sktime` pull requests!

What woul be great if you could add this to the API reference in `docs/source/api_reference/forecasting` next to the expanding window splitter, so users can find it easily!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks like it should work - the failure is due to the automated testing framework requiring a `get_test_params` function that returns valid parameter sets for parameters without a default.

There's currently no extension guide for splitters - this is tracked in issue https://github.com/sktime/sktime/issues/4889 - but you can have a look at any other extension guide, e.g., forecasters in `extension_templates/forecasting_simple`. I'd suggest we return 2 parameter sets, one where only `test_size` is set, and one where `folds` and `step_length` are additionally set to a non-default (while `test_size` is sth else than in the 1st set).

**Comment by fkiraly** [comment]
@davidgilbertson, I hope it's ok that I added the `get_test_params` - just wanted to see the test execution results.
Feel free to revert.

**Review by fkiraly — DISMISSED** [review]
I added the `get_test_params`, and it passes the tests now - so I'm fine with it, if you are fine with my changes.

Nice addition, should make "chop from the end" validation easier, especially if series are different lengths!

**Comment by davidgilbertson** [comment]
Sorry for the delay, PR updated, should be ready for review/merge now.

**Review by fkiraly — APPROVED** [review]
Thanks! Looks good.
